### ISSUE TITLE
intermediate tokens

### ISSUE THREAD

**Issue by wangyf8848** [facebookresearch/dinov2#398]
Hi，

I have got the mdoel using : 'dinov2_vitl14 = torch.hub.load('facebookresearch/dinov2', 'dinov2_vitl14')'.
And how can I get the output tokens of specific intermediate layers when inferencing.

**Comment by aeinsara** [comment]
hi, do you find a solution?

**Comment by paulgay** [comment]
x = torch.rand(1, 3, 224, 224)
x = dinov2_vitb14_reg_lc.backbone.patch_embed(x)
blocks = []
with torch.no_grad():
    for m in dinov2_vitb14_reg_lc.backbone.blocks:
        x = m(x) 
        blocks.append(x)

See also get_intermediate_layers function in dinov2/models/vision_transformer.py

**Comment by qasfb** [comment]
get_intermediate_layers is indeed what you need !

**Comment by Shar-01** [comment]
Can someone confirm what is the size of each of the different tokens? `['x_norm_clstoken', 'x_norm_regtokens', 'x_norm_patchtokens', 'x_prenorm', 'masks']`. I was considering the Huggingface Dinov2 and when I see the `last_hidden_state`, it returns an encoder output size of `(b, 257, 768)` where `257` are the num of tokens. I went to their [source code](https://github.com/huggingface/transformers/blob/v4.42.0/src/transformers/models/dinov2/modeling_dinov2.py#L569) and found they are returning `encoder_output[0]` as the final hidden state. When I see here that there are 4 different intermediate outputs, I wonder what their `encoder_output[0]` is. I will look deeper, meanwhile if someone already knows something, would be great to know.

**Comment by barbolo** [comment]
The output from the huggingface dinov2 model is probably `(b, 1, 768)` class token and `(b, 256, 768)` patch tokens. The input size is probably 224 (so you have 16 x 16 patch tokens, each one with patch size 14 x 14).

**Comment by Shar-01** [comment]
Thanks, @barbolo. Yes, this is how it looked when I did not see the Dinov2 intermediate features from here. But now, it is clear from the reply here: https://github.com/huggingface/transformers/issues/27977#issuecomment-2240080829
